src/cargo-fmt/main.rs

```diff
@@ -455,10 +512,70 @@ fn get_targets_with_hitlist(
     }
 }
 
-fn add_targets(target_paths: &[cargo_metadata::Target], targets: &mut BTreeSet<Target>) {
+fn add_targets(
+    manifest_path: &PathBuf,
+    target_paths: &[cargo_metadata::Target],
+    targets: &mut BTreeSet<Target>,
+    include_nested_test_files: bool,
+) -> Result<(), io::Error> {
+    let mut test_files_added = false;
     for target in target_paths {
-        targets.insert(Target::from_target(target));
+        // Packages often have more than one `test` target,
+        // so only add the nested files for the first one.
+        let check_for_nested_test_files = include_nested_test_files
+            && !test_files_added
+            && target.kind.iter().any(|t| t == "test");
+
+        if !check_for_nested_test_files {
+            targets.insert(Target::from_target(&target, None));
+            continue;
+        }
+
+        if let Some(package_dir) = manifest_path.parent() {
+            let target_dir = package_dir.join("tests");
+            test_files_added = true;
+            let test_files = get_nested_integration_test_files(&target_dir, &target_dir);
+            if test_files.is_none() {
+                return Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "Error encountered while searching for nested integration test files",
+                ));
+            }
+            targets.insert(Target::from_target(&target, test_files));
+        } else {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Unable to determine root `tests` directory for /
+                nested integration test file discovery",
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+// Returns a `Vec` containing `PathBuf`s of nested .rs files within subdirectories
+// of the `tests` directory for a given package.
+// https://github.com/rust-lang/rustfmt/issues/1820
+fn get_nested_integration_test_files(parent_path: &Path, root_dir: &Path) -> Option<Vec<PathBuf>> {
```

Comment:
Nit: the naming `parent_path` seems a bit odd, as it points to the actual target directory we are going to traverse.

Reply:
It does now that you mention it. I'll revert it back

Reply:
Done!